editor/config_loader.cpp

```diff
@@ -0,0 +1,203 @@
+#include "config_loader.hpp"
+
+#include "platform/platform.hpp"
+
+#include "coding/internal/file_data.hpp"
+#include "coding/reader.hpp"
+
+#include "base/logging.hpp"
+
+#include "std/fstream.hpp"
+
+#include "3party/pugixml/src/pugixml.hpp"
+#include "3party/Alohalytics/src/http_client.h"
+
+namespace
+{
+using alohalytics::HTTPClientPlatformWrapper;
+
+auto const * kConfigFileName = "editor.config";
+auto const * kHashFileName = "editor.config.hash";
+auto const kSynchroTimeout = seconds(3600);
+auto const * kRemoteHashUrl = "remote hash URL";
+auto const * kRemoteConfigUrl = "remote config URL";
+
+string GetConfigFilePath() { return GetPlatform().WritablePathForFile(kConfigFileName); }
+string GetHashFilePath() { return GetPlatform().WritablePathForFile(kHashFileName); }
+
+string RunSimpleHttpRequest(string const & url)
+{
+  HTTPClientPlatformWrapper request(url);
+  bool reqResult = request.RunHTTPRequest();
+
+  if (reqResult &&
+      !request.was_redirected() &&
+      request.error_code() == 200)  // 200 - http status OK
+  {
+    return request.server_response();
+  }
+
+  return {};
+}
+}  // namespace
+
+namespace editor
+{
+void Waiter::Interrupt()
+{
+  {
+    lock_guard<mutex> lock(m_mutex);
+    m_interrupted = true;
+  }
+
+  m_event.notify_all();
+}
+
+template <class _Rep, class _Period>
+bool Waiter::Wait(duration<_Rep, _Period> const & d)
+{
+  unique_lock<mutex> lock(m_mutex);
+
+  if(m_interrupted)
+    return false;
+
+  try
+  {
+    // Avoid spurious wakeups.
+    auto endTime = high_resolution_clock::now() + d;
+    while(!m_interrupted && endTime > high_resolution_clock::now())
+    {
+      m_event.wait_until(lock, endTime);
+    }
+  }
+  catch (RootException const & ex)
+  {
+    LOG(LERROR, (ex.Msg()));
+  }
+
+  return true;
+}
+
+ConfigLoader::ConfigLoader(pugi::xml_document & doc, mutex & mtx)
+  : m_docMutex(mtx)
+  , m_document(doc)
+{
+  LoadFromLocal();
+  m_loaderThr = thread(&ConfigLoader::LoadFromServer, this);
+}
+
+ConfigLoader::~ConfigLoader()
+{
+  m_waiter.Interrupt();
+
+  if (m_loaderThr.joinable())
+    m_loaderThr.join();
+}
+
+void ConfigLoader::LoadFromLocal()
+{
+  string content;
+  auto const reader = GetPlatform().GetReader(kConfigFileName);
+  reader->ReadAsString(content);
+
+  pugi::xml_document doc;
+  if (!doc.load_buffer(content.data(), content.size()))
+    MYTHROW(ConfigLoadError, ("Can't parse config"));
+
+  lock_guard <mutex> lock(m_docMutex);
+  m_document.reset(doc);
+}
+
+void ConfigLoader::LoadFromServer()
+{
+  auto const saveAndReload = [this](pugi::xml_document const & doc)
+  {
+    auto const filePath = GetConfigFilePath();
+    auto const result =
+      my::WriteToTempAndRenameToFile(filePath, [&doc](string const & fileName)
+      {
+        return doc.save_file(fileName.c_str(), " " /* indent */);
```

Comment:
You can remove indent entirely or set it as "". No one cares how it is stored on device.